pkg/catalog/catalog.go

```diff
@@ -161,21 +172,43 @@ func newSharedCatalog() (*Catalog, error) {
 	return c, nil
 }
 
-// getCatalogCache retrieves the catalog from the local directory.
-func getCatalogCache() (catalog *Catalog, err error) {
-	b, err := os.ReadFile(getCatalogCachePath())
+// getCatalogCache retrieves the catalog from the local directory along with locking the catalog file
+// If `setWriteLock` is false, it will read the catalog file with ReadLock and release the lock at the same time
+// If `setWriteLock` is true, it will apply WriteLock to the catalog file, read the catalog file
+// and keep the WriteLock to the file along with returning `unlock` function. It is caller
+// functions responsibility to unlock the WriteLock after the catalog update
+func getCatalogCache(setWriteLock bool) (catalog *Catalog, unlock func(), err error) {
+	unlock = func() {}
+	var b []byte
+
+	if setWriteLock {
+		var lockedFile *lockedfile.File
+		lockedFile, err = lockedfile.Edit(getCatalogCachePath())
+		if err != nil {
+			return nil, nil, err
+		}
+		unlock = func() {
+			if lockedFile != nil {
+				lockedFile.Close()
+			}
+		}
+		b, err = io.ReadAll(lockedFile)
+	} else {
+		b, err = lockedfile.Read(getCatalogCachePath())
+	}
+
 	if err != nil {
 		catalog, err = newSharedCatalog()
```

Comment:
It looks like that could still be a potential race if multiple callers enter the newSharedCatalog case (because there is no write lock acquired by either)

Reply:
It shouldn't be a problem because we will lock the file at `L185` or `L197`.

Note: `newSharedCatalog` creates a new in-memory catalog object and does not create a new file.